### PR TITLE
Stabilize GlobalTimeoutTest and EngineTimeoutTest

### PR DESCRIPTION
**Pull request by aSemy**
Wait until the engine has completed before asserting the results.

Part of #4113

### CODE REVIEW

File: kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/listener/CollectingTestEngineListener.kt [hunk 1]
```diff
@@ -16,31 +17,38 @@ class CollectingTestEngineListener : AbstractTestEngineListener(), Mutex by Mute
    val names = mutableListOf<String>()
    var errors = false
 
+   /**
+    * An active [Job] that will be completed when [engineFinished] is invoked.
+    * @see waitForEngineFinished
+    */
+   private val engineFinishedJob = Job()
```
Comment (OliverO2) [inline kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/listener/CollectingTestEngineListener.kt]: This is an interesting solution, using a Job as a boolean-like state which one can wait for while suspending.

It took me some time to decipher, as I had to make sure that the Job
* does not perform actual work, and
* does not have children (in which case `complete()` would not make `join()` return immediately).

Could you add some comment to that effect?

([The same could be achieved with a `Mutex` (playground)](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMi4wLjAiLCJwbGF0Zm9ybSI6ImphdmEiLCJhcmdzIjoiIiwibm9uZU1hcmtlcnMiOnRydWUsInRoZW1lIjoiaWRlYSIsImNvZGUiOiJpbXBvcnQga290bGlueC5jb3JvdXRpbmVzLmRlbGF5XG5pbXBvcnQga290bGlueC5jb3JvdXRpbmVzLmxhdW5jaFxuaW1wb3J0IGtvdGxpbnguY29yb3V0aW5lcy5ydW5CbG9ja2luZ1xuaW1wb3J0IGtvdGxpbnguY29yb3V0aW5lcy5zeW5jLk11dGV4XG5pbXBvcnQga290bGlueC5jb3JvdXRpbmVzLnN5bmMud2l0aExvY2tcbmltcG9ydCBrb3RsaW4udGltZS5EdXJhdGlvbi5Db21wYW5pb24uc2Vjb25kc1xuXG5mdW4gbWFpbigpOiBVbml0ID0gcnVuQmxvY2tpbmcge1xuICAgIHZhbCBlbmdpbmVGaW5pc2hlZCA9IE11dGV4KGxvY2tlZCA9IHRydWUpXG5cbiAgICBsYXVuY2gge1xuICAgICAgICBkZWxheSgyLnNlY29uZHMpXG4gICAgICAgIGVuZ2luZUZpbmlzaGVkLnVubG9jaygpXG4gICAgICAgIHByaW50bG4oXCJmaW5pc2hlZFwiKVxuICAgIH1cblxuICAgIGxhdW5jaCB7XG4gICAgICAgIGRlbGF5KDAuNS5zZWNvbmRzKVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMTogd2FpdGluZ1wiKVxuICAgICAgICBlbmdpbmVGaW5pc2hlZC53aXRoTG9jayB7fVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMTogZGV0ZWN0ZWQgZmluaXNoXCIpXG4gICAgfVxuXG4gICAgbGF1bmNoIHtcbiAgICAgICAgZGVsYXkoMS5zZWNvbmRzKVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMjogd2FpdGluZ1wiKVxuICAgICAgICBlbmdpbmVGaW5pc2hlZC53aXRoTG9jayB7fVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMjogZGV0ZWN0ZWQgZmluaXNoXCIpXG4gICAgfVxuXG4gICAgbGF1bmNoIHtcbiAgICAgICAgZGVsYXkoMy5zZWNvbmRzKVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMzogd2FpdGluZ1wiKVxuICAgICAgICBlbmdpbmVGaW5pc2hlZC53aXRoTG9jayB7fVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMzogZGV0ZWN0ZWQgZmluaXNoXCIpXG4gICAgfVxufVxuIn0=), but that would also require some comment explaining to its uncommon use. I have a similar thing called [`Gate` in this KotlinConf talk companion repo](https://github.com/OliverO2/kmp-e2e-stress-testing-example/blob/main/application/library-testing/src/commonMain/kotlin/library/testing/Gate.kt).)

**Review by OliverO2 — CHANGES_REQUESTED** [review]

File: kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/listener/CollectingTestEngineListener.kt (hunk 1 above)
Comment (aSemy) [reply kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/listener/CollectingTestEngineListener.kt]: Ahh yes, the `mutex.withLock {}` is probably better. I'll update the code and with a comment.

**Review by OliverO2 — APPROVED** [review]

**Comment by aSemy** [comment]
Hmmm apparently `engineFinished()` can get called multiple times?

```
> Task :kotest-framework:kotest-framework-engine:jvmTest

com.sksamuel.kotest.engine.extensions.spec.BeforeSpecListenerTest[jvm]
  Test BeforeSpecListener inline should be triggered before tests[jvm] FAILED (2.3s)
  java.lang.IllegalStateException: This mutex is not locked
  Caused by: java.lang.IllegalStateException: This mutex is not locked

FAILURE: Executed 1423 tests in 6m 5s (1 failed, 186 skipped)
```

File: kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/listener/CollectingTestEngineListener.kt [hunk 2]
```diff
@@ -16,31 +19,44 @@ class CollectingTestEngineListener : AbstractTestEngineListener(), Mutex by Mute
    val names = mutableListOf<String>()
    var errors = false
 
+   /**
+    * Counts the number of times [engineStarted] and [engineFinished] have been invoked.
+    *
+    * Used by [waitForEngineFinished], so that internal Kotest tests can wait until engines
+    * are finished before performing assertions.
+    */
+   private val activeEngines = MutableStateFlow<Int?>(null)
```
Comment (OliverO2) [inline kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/listener/CollectingTestEngineListener.kt]:
```suggestion
   private val activeEngineCount = MutableStateFlow<Int?>(null)
```

`activeEngines` suggests some collection of `Engine`s. I find that in such cases `count` helps at use sites.

File: kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/listener/CollectingTestEngineListener.kt [hunk 3]
```diff
@@ -56,4 +72,10 @@ class CollectingTestEngineListener : AbstractTestEngineListener(), Mutex by Mute
    fun TestCase.toKey(): TestCaseKey {
       return TestCaseKey(this.descriptor, this.name, this.spec::class)
    }
+
+   /** Suspends until at least one engine has started, and all engines have finished. */
+   internal suspend fun waitForEngineFinished() {
```
Comment (OliverO2) [inline kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/listener/CollectingTestEngineListener.kt]:
```suggestion
   internal suspend fun waitForEnginesFinished() {
```

**Review by OliverO2 — APPROVED** [review]
Looks good, given the new insights. I also wonder why the same listener would observe multiple engines. 🤔

**Comment by aSemy** [comment]
> Looks good, given the new insights. I also wonder why the same listener would observe multiple engines. 🤔

Oh, I figured it out and forgot to update my message. The reason is one of the test re-uses a listener, which I think is reasonable.

https://github.com/kotest/kotest/blob/e2ebd497ee304128ae2bf7325505565b3263b73c/kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/extensions/spec/BeforeSpecListenerTest.kt#L78-L90

**Comment by OliverO2** [comment]
Thanks for clarifying! Helps to demystify what I suspected was one more piece of magic. Now I can relax again. :)

**Pull request merged**